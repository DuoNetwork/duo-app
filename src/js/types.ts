export interface IPriceDatum {
	date: string;
	ETH: number;
	ClassA: number;
	ClassB: number;
	ClassAbeforeReset?: number;
	ClassBbeforeReset?: number;
	ResetType?: string;
	InterestCount: number;
}

export type IPriceData = IPriceDatum[];

export interface IMVDatum {
	date: string;
	MV: number;
}

export type IMVData = IMVDatum[];

export interface IPriceProps {
	name: string;
	data: IPriceData;
	movedata: IPriceData;
	pickedPriceDatum: (d: IPriceDatum) => void;
}

export interface IPriceState {
	windowWidth: number;
	windowHeight: number;
}

export interface IMVProps {
	name: string;
	data: IMVData;
	pickedMVDatum: (d: IMVDatum) => void;
}

export interface IMVState {
	windowWidth: number;
	windowHeight: number;
}

export interface IAssets {
	ETH: number;
	ClassA: number;
	ClassB: number;
}
