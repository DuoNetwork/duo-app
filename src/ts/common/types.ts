import {
	ICustodianAddresses,
	IDualClassStates,
	IEsplanadeStates,
	IMagiStates,
	IVotingData
} from '@finbook/duo-contract-wrapper';
import { IAcceptedPrice, IConversion, IPrice, IStatus } from '@finbook/duo-market-data';
import { AnyAction } from 'redux';
import { ThunkAction } from 'redux-thunk';

export interface ISourceData<T> {
	// bitfinex: T;
	gemini: T;
	kraken: T;
	gdax: T;
	[source: string]: T;
}

export interface IState {
	readonly dualClass: IDualClassState;
	readonly magi: IMagiState;
	readonly esplanade: IEsplanadeState;
	readonly web3: IWeb3State;
	readonly dynamo: IDynamoState;
	readonly ui: IUIState;
}

export interface IWeb3State {
	readonly account: string;
	readonly network: number;
	readonly gasPrice: number;
	readonly balance: number;
}

export interface IDualClassState {
	readonly type: string;
	readonly tenor: string;
	readonly subscription: number;
	readonly states: IDualClassStates;
	readonly addresses: ICustodianAddresses;
	readonly exchangePrices: IPrice[];
	readonly acceptedPrices: IAcceptedPrice[];
	readonly conversions: IConversion[];
	readonly balances: {
		a: number;
		b: number;
	};
}

export interface IEsplanadeState {
	readonly subscription: number;
	readonly states: IEsplanadeStates;
	readonly moderator: {
		address: string;
		balance: number;
	};
	readonly candidate: {
		address: string;
		balance: number;
	};
	readonly coldAddressPool: IAddresses;
	readonly hotAddressPool: IAddresses;
	readonly custodianPool: IAddresses;
	readonly otherContractPool: IAddresses;
	readonly votingData: IVotingData;
}
export interface IAddresses {
	[address: string]: {
		index: number;
		balance?: number;
		label?: string;
	};
}

export interface IMagiState {
	readonly subscription: number;
	readonly states: IMagiStates;
	readonly operator: IAccount;
	readonly roleManager: IAccount;
	readonly priceFeeds: IMagiPriceFeed;
	readonly acceptedPrices: IAcceptedPrice[];
}

export interface IAccount {
	balance: number;
	address: string;
}

export interface IMagiPriceFeed {
	[address: string]: {
		balance: number;
		index: number;
	};
}

export interface IDynamoState {
	readonly status: IStatus[];
}

export interface IUIState {
	readonly period: number;
	readonly source: string;
	readonly locale: string;
}

export type VoidThunkAction = ThunkAction<void, IState, undefined, AnyAction>;

export interface ITableRecord {
	[key: string]: any;
}

export enum WsChannelMessageTypes {
	Add = 'add',
	Update = 'update',
	Cancel = 'cancel',
	Subscribe = 'subscribe'
}

export enum WsChannelName {
	Orderbook = 'orderbook',
	Order = 'order'
}
