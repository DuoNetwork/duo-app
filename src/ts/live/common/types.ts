import { AnyAction } from 'redux';
import { ThunkAction } from 'redux-thunk';
export * from '../../../../../duo-admin/src/common/types';
import * as adminTypes from '../../../../../duo-admin/src/common/types';

export interface ICustodianPrice {
	address: string;
	price: number;
	timestamp: number;
}

export interface ICustodianStates {
	state: string;
	navA: number;
	navB: number;
	totalSupplyA: number;
	totalSupplyB: number;
	alpha: number;
	beta: number;
	feeAccumulated: number;
	periodCoupon: number;
	limitPeriodic: number;
	limitUpper: number;
	limitLower: number;
	createCommRate: number;
	redeemCommRate: number;
	period: number;
	iterationGasThreshold: number;
	ethDuoFeeRatio: number;
	preResetWaitingBlocks: number;
	priceTol: number;
	priceFeedTol: number;
	priceFeedTimeTol: number;
	priceUpdateCoolDown: number;
	numOfPrices: number;
	nextResetAddrIndex: number;
	lastAdminTime: number;
	adminCoolDown: number;
	usersLength: number;
	addrPoolLength: number;
	ethBalance: number;
	duoBalance: number;
}

export interface ICustodianPrices {
	first: ICustodianPrice;
	second: ICustodianPrice;
	reset: ICustodianPrice;
	last: ICustodianPrice;
}

export interface IBalances {
	eth: number;
	duo: number;
	allowance: number;
	tokenA: number;
	tokenB: number;
}

export interface IAccountBalances extends IBalances {
	account: string;
}

export interface IAddress {
	address: string;
	balance: number;
}

export interface IAddresses {
	operator: IAddress;
	feeCollector: IAddress;
	priceFeed1: IAddress;
	priceFeed2: IAddress;
	priceFeed3: IAddress;
	poolManager: IAddress;
	[role: string]: IAddress;
}

export interface ISourceData<T> {
	bitfinex: T;
	gemini: T;
	kraken: T;
	gdax: T;
	[source: string]: T;
}

export interface IState {
	readonly contract: IContractState;
	readonly dynamo: IDynamoState;
	readonly ui: IUIState;
	readonly ws: IWSState;
}

export interface IContractState {
	readonly states: ICustodianStates;
	readonly prices: ICustodianPrices;
	readonly balances: IBalances;
	readonly addresses: IAddresses;
	readonly addressPool: IAddress[];
	readonly account: string;
	readonly network: number;
	readonly gasPrice: number;
	readonly allBalances: {[index: number]: IAccountBalances}
}

export interface IDynamoState {
	readonly status: adminTypes.IStatus[];
	readonly hourly: ISourceData<adminTypes.IPrice[]>;
	readonly minutely: ISourceData<adminTypes.IPrice[]>;
	readonly price: adminTypes.IAcceptedPrice[];
	readonly last: ISourceData<ICustodianPrice>;
	readonly conversion: adminTypes.IConversion[];
	readonly totalSupply: adminTypes.ITotalSupply[]
}

export interface IUIState {
	readonly refresh: number;
	readonly locale: string;
}

export type VoidThunkAction = ThunkAction<void, IState, undefined, AnyAction>;

export interface ITableRecord {
	[key: string]: any
}

export interface IWSOrderBook extends IWSChannel {
	changes: [
		{
			side: string;
			price: string;
			amount: string;
		}
	];
}

export interface IWSOrderBookChanges {
	side: string;
	price: string;
	amount: string;
}

export interface IWSChannel {
	type: string;
	channel: {
		name: string;
		marketId: string;
	};
	requestId: number;
	timestamp: number;
	delay?: number;
}

export interface IWSOrderBookSubscription extends IWSChannel {
	bids: [
		{
			makerTokenName: string;
			takerTokenName: string;
			marketId: string;
			side: string;
			amount: number;
			price: number;
		}
	];
	asks: [
		{
			makerTokenName: string;
			takerTokenName: string;
			marketId: string;
			side: string;
			amount: number;
			price: number;
		}
	];
}

export enum WsChannelMessageTypes {
	Add = 'add',
	Update = 'update',
	Cancel = 'cancel',
	Subscribe = 'subscribe'
}

export enum WsChannelName {
	Orderbook = 'orderbook',
	Order = 'order'
}

export interface IWSState {
	readonly subscribe: IWSOrderBookSubscription;
}
