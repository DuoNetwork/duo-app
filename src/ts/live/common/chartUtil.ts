import moment from 'moment';
import * as CST from '../common/constants';
import { IAcceptedPrice, ICustodianPrice, ICustodianStates, IPriceBar, ISourceData } from './types';

class ChartUtil {
	public interpolate(sourceData: IPriceBar[], isHourly: boolean): IPriceBar[] {
		if (!sourceData.length) return [];
		sourceData.sort((a, b) => a.timestamp - b.timestamp);
		const newSourceData: IPriceBar[] = [];
		const timeStep = isHourly ? 3.6e6 : 60000;
		const source = sourceData[0].source;
		for (let i = 0; i < sourceData.length - 1; i++) {
			newSourceData.push(sourceData[i]);
			const timeIndex = sourceData[i].timestamp;
			const nextTimeIndex = sourceData[i + 1].timestamp;
			const close = sourceData[i].close;
			let newTimeStamp = timeIndex + timeStep;
			while (newTimeStamp < nextTimeIndex) {
				newSourceData.push({
					source: source,
					date: moment(newTimeStamp).format('YYYY-MM-DD'),
					hour: moment(newTimeStamp).format('HH'),
					minute: isHourly ? 0 : Number(moment(newTimeStamp).format('mm')),
					open: close,
					high: close,
					low: close,
					close: close,
					volume: 0,
					timestamp: newTimeStamp
				});
				newTimeStamp += timeStep;
			}
		}
		newSourceData.push(sourceData[sourceData.length - 1]);

		return newSourceData;
	}

	public reset(
		price: IAcceptedPrice[],
		limitUp: number,
		limitDown: number,
		limitPeriod: number
	): IAcceptedPrice[] {
		return price
			.filter(p => p.navB >= limitUp || p.navB <= limitDown || p.navA >= limitPeriod)
			.map(p => ({
				price: p.price,
				navA: 1,
				navB: 1,
				timestamp: p.timestamp + 717
			}));
	}

	public mergeReset(prices: IAcceptedPrice[], resets: IAcceptedPrice[]): IAcceptedPrice[] {
		const all = [...prices, ...resets];
		all.sort((a, b) => a.timestamp - b.timestamp);
		return all;
	}

	public mergeLastToPriceBar(
		pricebars: ISourceData<IPriceBar[]>,
		last: ISourceData<ICustodianPrice>,
		isHourly: boolean
	) {
		CST.EXCHANGES.forEach(src => {
			const pricebar: IPriceBar[] = pricebars[src.toLowerCase()];
			if (!pricebar || !pricebar.length) return;
			const sourceLast: ICustodianPrice = last[src.toLowerCase()];
			const lastBar = pricebar[pricebar.length - 1];
			const dateObj = moment.utc(sourceLast.timestamp);
			if (isHourly && lastBar.date + ' ' + lastBar.hour !== dateObj.format('YYYY-MM-DD HH'))
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: 0,
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else if (
				!isHourly &&
				lastBar.date + ' ' + lastBar.hour + ' ' + lastBar.minute !==
					dateObj.format('YYYY-MM-DD HH m')
			)
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: Number(dateObj.format('mm')),
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else {
				lastBar.high = Math.max(lastBar.high, sourceLast.price);
				lastBar.low = Math.min(lastBar.low, sourceLast.price);
				lastBar.close = sourceLast.price;
				lastBar.timestamp = sourceLast.timestamp;
			}
		});
	}

	public mergeLastToPrice(
		prices: IAcceptedPrice[],
		states: ICustodianStates,
		last: ICustodianPrice
	) {
		if (!prices.length) return;
		const lastPrice = prices[prices.length - 1];
		if (lastPrice.timestamp < last.timestamp)
			prices.push({
				price: last.price,
				navA: states.navA,
				navB: states.navB,
				timestamp: last.timestamp
			});
	}
}

const chartUtil = new ChartUtil();
export default chartUtil;
