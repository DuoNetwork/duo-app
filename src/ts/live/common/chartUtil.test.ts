import moment from 'moment';
import chartUtil from './chartUtil';
import { IPriceBar } from './types';

const hourlyBar1 = {
	source: 'source',
	date: '2018-06-06',
	hour: '00',
	minute: 0,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
};

const hourlyBar2 = {
	source: 'source',
	date: '2018-06-06',
	hour: '01',
	minute: 0,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 01:0', 'YYYY-MM-DD HH:m').valueOf()
};

const hourlyBar3 = {
	source: 'source',
	date: '2018-06-06',
	hour: '03',
	minute: 0,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 03:0', 'YYYY-MM-DD HH:m').valueOf()
};

test('does not interpolate unnecessarily', () => {
	const result = chartUtil.interpolate([hourlyBar1, hourlyBar2], true);
	expect(result.length).toBe(2);
	expect(result[0]).toBe(hourlyBar1);
	expect(result[1]).toEqual(hourlyBar2);
});

test('interpolate correctly', () => {
	const result = chartUtil.interpolate([hourlyBar1, hourlyBar3], true);
	expect(result.length).toBe(4);
	expect(result[0]).toBe(hourlyBar1);
	expect(result[3]).toEqual(hourlyBar3);
	expect(result[1]).toMatchSnapshot();
	expect(result[2]).toMatchSnapshot();
});

const minutelyBar1 = {
	source: 'source',
	date: '2018-06-06',
	hour: '00',
	minute: 0,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
};

const minutelyBar2 = {
	source: 'source',
	date: '2018-06-06',
	hour: '00',
	minute: 1,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 00:1', 'YYYY-MM-DD HH:m').valueOf()
};

const minutelyBar3 = {
	source: 'source',
	date: '2018-06-06',
	hour: '00',
	minute: 3,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment('2018-06-06 00:3', 'YYYY-MM-DD HH:m').valueOf()
};

test('does not interpolate unnecessarily', () => {
	const result = chartUtil.interpolate([minutelyBar1, minutelyBar2], false);
	expect(result.length).toBe(2);
	expect(result[0]).toBe(minutelyBar1);
	expect(result[1]).toEqual(minutelyBar2);
});

test('interpolate correctly', () => {
	const result = chartUtil.interpolate([minutelyBar1, minutelyBar3], false);
	expect(result.length).toBe(4);
	expect(result[0]).toBe(minutelyBar1);
	expect(result[3]).toEqual(minutelyBar3);
	expect(result[1]).toMatchSnapshot();
	expect(result[2]).toMatchSnapshot();
});
