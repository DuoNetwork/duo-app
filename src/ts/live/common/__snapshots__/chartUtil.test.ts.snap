// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find reset points correctly 1`] = `
Array [
  Object {
    "blockNumber": 222,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 2717,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 3717,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 4717,
    "transactionHash": "ddd",
  },
]
`;

exports[`hourly interpolate correctly 1`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 2,
  "low": 2,
  "open": 2,
  "period": 60,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528246800000,
  "volume": 0,
}
`;

exports[`hourly interpolate correctly 2`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 2,
  "low": 2,
  "open": 2,
  "period": 60,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528250400000,
  "volume": 0,
}
`;

exports[`hourly interpolate correctly 3`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 3,
  "low": 0,
  "open": 1,
  "period": 60,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528243200000,
  "volume": 246,
}
`;

exports[`merge reset correctly 1`] = `
Array [
  Object {
    "blockNumber": 111,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 1000,
    "transactionHash": "aaa",
  },
  Object {
    "blockNumber": 222,
    "navA": 1,
    "navB": 2,
    "price": 123,
    "timestamp": 2000,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 222,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 2717,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "navA": 1,
    "navB": 0.25,
    "price": 123,
    "timestamp": 3000,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 333,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 3717,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "navA": 1.1,
    "navB": 1,
    "price": 123,
    "timestamp": 4000,
    "transactionHash": "ddd",
  },
  Object {
    "blockNumber": 444,
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 4717,
    "transactionHash": "ddd",
  },
]
`;

exports[`merge total supply correctly 1`] = `
Array [
  Object {
    "blockNumber": 111,
    "timestamp": 1234567890,
    "tokenA": 123,
    "tokenB": 456,
    "transactionHash": "abc",
  },
  Object {
    "blockNumber": 0,
    "timestamp": 2345678901,
    "tokenA": 1234,
    "tokenB": 2345,
    "transactionHash": "",
  },
]
`;

exports[`minutely interpolate correctly 1`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 2,
  "low": 2,
  "open": 2,
  "period": 1,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528243260000,
  "volume": 0,
}
`;

exports[`minutely interpolate correctly 2`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 2,
  "low": 2,
  "open": 2,
  "period": 1,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528243320000,
  "volume": 0,
}
`;

exports[`minutely interpolate correctly 3`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 3,
  "low": 0,
  "open": 1,
  "period": 1,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528243200000,
  "volume": 246,
}
`;
